server/src/parser/operators.ts

```diff
@@ -0,0 +1,76 @@
+
+// Return the precedence priority of the operator.
+//
+// As unary `-` can't be distinguished from binary `-`, unary precedence
+// isn't described in this table but in `Parser.parseSubExpression()` itself.
+//
+// As this gets hit on every expression it's been optimized due to
+// the expensive CompareICStub which took ~8% of the parse time.
+
+export default class Operators {
+
+  static fakeMaxPrecedenceOperator = ';';
+  static maxPrecedenceValue = 99;
+  static minPrecedenceValue = 0;
+
+  static binaryPrecedenceOf(operator: string): number {
+    const charCode = operator.charCodeAt(0);
+    const length = operator.length;
+
+    if (1 === length) {
+      switch (charCode) {
+      case 59: return Operators.maxPrecedenceValue; // # (fake one, a "guard")
+      case 94: return 12; // ^
+      case 42: case 47: case 37: case 92: return 10; // * / % \
+      case 43: case 45: return 9; // + -
+      case 38: return 6; // & (bitwise AND)
+      case 124: return 4; // | (bitwise OR)
+      case 60: case 62: return 3; // < >
+      }
+    } else if (2 === length) {
+      switch (charCode) {
+      case 46: return 8; // ..
+      case 60: case 62:
+        if('<<' === operator || '>>' === operator) {
+          return 7;
+        } // << >>
+        return 3; // <= >=
+      case 33: case 61: case 126: return 3; // == ~= !=
+      case 111: return 1; // or
+      case 94: return 5; // ^^ (bitwise XOR, pico-8 lua uses the normal bitwise XOR ~ as bitwise NOT)
+      }
+    } else if (3 === length) {
+      switch (operator) {
+      case '>>>': case '<<>': case '>><': return 7;
+      case 'and': return 2;
+      }
+    }
+    return Operators.minPrecedenceValue;
+  }
+
+  static isAssociative(operator: string): boolean {
```

Comment:
Typically operators are referred to as `left-associative` or `right-associative`, rather than just `associative` -- I think what this function is returning is whether the operator is left-associative. Mind renaming it to reflect that?

Reply:
Disclaimer: I am not an expert in the field, I only rely on my quick googling 🙂 

In this case left or right would be incorrect, because `*` and `+` (some operators this function returns `true` for) are nor left nor right associative. They are "just" associative. 

See this [excerpt from Wikipedia](https://en.wikipedia.org/wiki/Operator_associativity):

> Operators may be **associative** (meaning the operations can be grouped arbitrarily), **left-associative** (meaning the operations are grouped from the left), **right-associative** (meaning the operations are grouped from the right) or **non-associative** (meaning operations cannot be chained, often because the output type is incompatible with the input types).

And another one:

> Addition and multiplication, by contrast, are **both left and right associative**. (e.g. `(a * b) * c = a * (b * c)`).

I see that the naming might be misleading. WDYT about naming it "both left and right associative". I know the name will become really lengthy, but way less ambiguous. I added a commit with it.

Reply:
That is talking about math in general -- it's a bit different for parsing programming languages specifically. For parsing, it's important that we strictly define which direction an operator associates in.

More info:
- http://craftinginterpreters.com/parsing-expressions.html
- https://blog.jcoglan.com/2017/07/07/precedence-and-associativity-in-recursive-descent/

In lua, [according to this page](https://www.lua.org/pil/3.5.html):

> All binary operators are left associative, except for `^´ (exponentiation) and `..´ (concatenation), which are right associative.

Probably the best way to handle this is that this function can just be called `getAssociativity` or something, and return either `"left"`, `"right"`, or `null`, or define an enum `Associativity` with the 3 options as numbers and return one of those (similar to `boundsCompare()`).

Reply:
I see myself struggling to wrap head around it 🙁 According to [Lua docs you linked](https://www.lua.org/pil/3.5.html) both `-` and `+` are **left** associative. If I have a function which returns same `LeftAssociativity` for both, it wouldn't work for the particular case I use the function for – in the case I solve, I need to distinguish between `-` and `+`, since `a + (b + c)` can simplify to `a + b + c` while `a - (b - c)` does not.

To remind my usage, in a simplified way:

```js
if (isRightSideOfAnExpression && !Operators.isBothLeftAndRightAssociative(parentOperator)) {
  return `(${expression})`;
} else {
  return expression;
}
```

I am not sure if the strict associativity for programming language applies here, since we do not talk here about parsing of a written expression, but about formatting already parsed one (in form of AST) into possible simplest representation. But again, I do not feel solid enough in the area to have a strong opinion, just regular gut feeling.

---

At this point I see 2 reasonable paths forwards:
a) you add commit with changes to my `Operator.isBothLeftAndRightAssociative` the way you see it correct (and make sure `removes parentheses from calculations when they are unnecessary` test passes green). It would make more sense since it seems that you know what you talk about and you know what do you want there, while I just try to interpret your expectations 🙂 
b) I rename `Operators.isBothLeftAndRightAssociative(operator)` to `Operators.doesNeedParenthesesIfOnTheRightSide(operator)` and invert boolean results and invert negation in the place I use it 😛 I like this solution because it doesn't try to operate on correct terms, just clearly states what it is for, but I also dislike this solution, because it seems right to me to operate on those math/programming terms when within its own `Operators` class 🤔 

To make it simpler in terms of processing forwards, I already did (b) in commit 1f77390bd5c75543142b5368fe2d1b16c39e312e – and feel free to rename function back or to any other one or rewrite it completely if going with (a) 🙂 👍